src/libraries/System.Net.Http/src/System/Net/Http/BrowserHttpHandler/BrowserHttpHandler.cs

```diff
@@ -309,77 +321,75 @@ private static HttpResponseMessage ConvertResponse(HttpRequestMessage request, W
-#if FEATURE_WASM_THREADS
-            });
-#endif
+                ObjectDisposedException.ThrowIf(_isDisposed, this);
+            } //lock
+        }
 
-            static async Task<HttpResponseMessage> Impl(HttpRequestMessage request, CancellationToken cancellationToken, bool? allowAutoRedirect)
+        public void Dispose()
+        {
+            lock (this)
             {
-                WasmFetchResponse fetchRespose = await CallFetch(request, cancellationToken, allowAutoRedirect).ConfigureAwait(true);
-                return ConvertResponse(request, fetchRespose);
+                if (_isDisposed)
+                    return;
+                _isDisposed = true;
+            }
+            _abortRegistration.Dispose();
+            if (_jsController != null)
+            {
+                if (!_jsController.IsDisposed)
+                {
+                    BrowserHttpInterop.AbortRequest(_jsController);// aborts also response
+                }
+                _jsController.Dispose();
             }
         }
     }
 
-    internal sealed class WasmHttpWriteStream : Stream
+    internal sealed class BrowserHttpWriteStream : Stream
     {
-        private readonly JSObject _transformStream;
-
-        public WasmHttpWriteStream(JSObject transformStream)
+        private readonly BrowserHttpController _controller; // we don't own it, we don't dispose it from here
+        public BrowserHttpWriteStream(BrowserHttpController controller)
         {
-            ArgumentNullException.ThrowIfNull(transformStream);
+            ArgumentNullException.ThrowIfNull(controller);
 
-            _transformStream = transformStream;
+            _controller = controller;
         }
 
         private Task WriteAsyncCore(ReadOnlyMemory<byte> buffer, CancellationToken cancellationToken)
         {
             cancellationToken.ThrowIfCancellationRequested();
-#if FEATURE_WASM_THREADS
-            return _transformStream.SynchronizationContext.Post(() => Impl(this, buffer, cancellationToken));
-#else
-            return Impl(this, buffer, cancellationToken);
-#endif
-            static async Task Impl(WasmHttpWriteStream self, ReadOnlyMemory<byte> buffer, CancellationToken cancellationToken)
+
+            MemoryHandle pinBuffer = buffer.Pin();
+            try
             {
-                using (Buffers.MemoryHandle handle = buffer.Pin())
-                {
-                    Task writePromise = TransformStreamWriteUnsafe(self._transformStream, buffer, handle);
-                    await BrowserHttpInterop.CancelationHelper(writePromise, cancellationToken).ConfigureAwait(true);
-                }
+                Task writePromise = BrowserHttpInterop.TransformStreamWriteUnsafe(_controller._jsController, buffer, pinBuffer);
+                return BrowserHttpInterop.CancellationHelper(writePromise, cancellationToken, _controller._jsController);
+            }
+            catch
+            {
+                pinBuffer.Dispose();
```

Comment:
This now only disposes the pin when an exception occurs, is that correct?

Reply:
fixed, thanks